storagenode/pieces/readwrite.go

```diff
@@ -32,28 +88,84 @@ func NewWriter(blob storage.BlobWriter) (*Writer, error) {
 // Write writes data to the blob and calculates the hash.
 func (w *Writer) Write(data []byte) (int, error) {
 	n, err := w.blob.Write(data)
-	w.size += int64(n)
+	w.pieceSize += int64(n)
 	_, _ = w.hash.Write(data[:n]) // guaranteed not to return an error
 	if err == io.EOF {
 		return n, err
 	}
 	return n, Error.Wrap(err)
 }
 
-// Size returns the amount of data written so far.
-func (w *Writer) Size() int64 { return w.size }
+// Size returns the amount of data written to the piece so far, not including the size of
+// the piece header.
+func (w *Writer) Size() int64 { return w.pieceSize }
 
 // Hash returns the hash of data written so far.
 func (w *Writer) Hash() []byte { return w.hash.Sum(nil) }
 
 // Commit commits piece to permanent storage.
-func (w *Writer) Commit(ctx context.Context) (err error) {
+func (w *Writer) Commit(ctx context.Context, pieceHeader *pb.PieceHeader) (err error) {
 	defer mon.Task()(&ctx)(&err)
 	if w.closed {
 		return Error.New("already closed")
 	}
+
+	// point of no return: after this we definitely either commit or cancel
 	w.closed = true
-	return Error.Wrap(w.blob.Commit(ctx))
+	defer func() {
+		if err != nil {
+			err = Error.Wrap(errs.Combine(err, w.blob.Cancel(ctx)))
+		} else {
+			err = Error.Wrap(w.blob.Commit(ctx))
+		}
+	}()
+
+	formatVer := w.blob.GetStorageFormatVersion()
+	pieceHeader.FormatVersion = int32(formatVer)
+	if formatVer == storage.FormatV0 {
```

Comment:
this check can happen one line above, before you assign it to `pieceHeader.FormatVersion`